File: ts/141.linked-list-cycle.ts
/*
 * @lc app=leetcode id=141 lang=typescript
 *
 * [141] Linked List Cycle
 */

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// @lc code=start

function hasCycle(head: ListNode | null): boolean {
  const nodesSeen = new Set<ListNode>();

  while (head !== null) {
    if (nodesSeen.has(head)) return true;
    nodesSeen.add(head);

    head = head.next;
  }

  return false;
}
// @lc code=end

File: go/141.linked-list-cycle/141.linked-list-cycle.go
/*
 * @lc app=leetcode id=141 lang=golang
 *
 * [141] Linked List Cycle
 */

package leetcode

import "github.com/maguroid/leetcode/go/structures"

type ListNode = structures.ListNode

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

/* using Floyd's Circle Finding Algorithm - Space Complexity O(1) */
func hasCycle(head *ListNode) bool {
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			return true
		}
	}

	return false
}

// @lc code=end

/* using Hash Table - Space Complexity O(n) */
func _(head *ListNode) bool {
	nodeMap := make(map[*ListNode]bool)

	for head != nil {
		if nodeMap[head] {
			return true
		}
		nodeMap[head] = true
		head = head.Next
	}

	return false
}
